google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java

```diff
@@ -442,22 +591,163 @@
+                return ApiFutures.immediateFuture(result);
+              }
+            },
+            MoreExecutors.directExecutor());
+
+    writeResultApiFuture.addListener(
+        new Runnable() {
+          public void run() {
+            pendingOperations.remove(operationCompletedFuture);
+            operationCompletedFuture.set(null);
+          }
+        },
+        MoreExecutors.directExecutor());
+
+    pendingOperations.add(operationCompletedFuture);
+    return writeResultApiFuture;
+  }
+
+  /** Used to schedule the write enqueue logic onto the bulkWriterExecutor. */
+  private ApiFuture<WriteResult> enqueueWrite(
+      final DocumentReference documentReference,
+      final OperationType operationType,
+      final BulkWriterOperationCallback operationCallback) {
+    return ApiFutures.transformAsync(
+        ApiFutures.immediateFuture(null),
+        new ApiAsyncFunction<Object, WriteResult>() {
+          public ApiFuture<WriteResult> apply(Object o) throws Exception {
+            return enqueueWriteHelper(documentReference, operationType, operationCallback, 0);
+          }
+        },
+        bulkWriterExecutor);
+  }
+
+  /**
+   * Adds the write to the appropriate batch queue and performs the bulkCommit. This helper method
+   * also catches failures and applies the user error callback and retrying if necessary.
+   */
+  private ApiFuture<WriteResult> enqueueWriteHelper(
+      final DocumentReference documentReference,
+      final OperationType operationType,
+      final BulkWriterOperationCallback operationCallback,
+      final int failedAttempts) {
+    List<BulkCommitBatch> operationBatchQueue = failedAttempts > 0 ? retryBatchQueue : batchQueue;
+    BulkCommitBatch bulkCommitBatch = getEligibleBatch(documentReference, operationBatchQueue);
+
+    // Send ready batches if this is the first attempt. Subsequent retry batches are scheduled
+    // after the initial batch returns.
+    if (failedAttempts == 0) {
+      sendReadyBatches(operationBatchQueue);
+    }
+
+    return ApiFutures.catchingAsync(
+        operationCallback.apply(bulkCommitBatch),
+        FirestoreException.class,
+        new ApiAsyncFunction<FirestoreException, WriteResult>() {
+          public ApiFuture<WriteResult> apply(FirestoreException exception)
+              throws BulkWriterError, ExecutionException, InterruptedException {
+            BulkWriterError bulkWriterError =
+                new BulkWriterError(
+                    exception.getStatus(),
+                    exception.getMessage(),
+                    documentReference,
+                    operationType,
+                    failedAttempts);
+
+            boolean shouldRetry = invokeUserErrorCallback(bulkWriterError).get();
+            logger.log(
+                Level.INFO,
+                String.format(
+                    "Running error callback on error code: %d, shouldRetry: %b",
```

Comment:
I think this log message would be more useful if it included the DocumentReference. It is also printed after the callback ran - so ideally, we would rephrase it a bit.